core/src/test/scala/org/apache/spark/scheduler/BarrierTaskContextSuite.scala

```diff
@@ -0,0 +1,148 @@
+    val conf = new SparkConf()
+      .setMaster("local-cluster[4, 1, 1024]")
+      .setAppName("test-cluster")
+    sc = new SparkContext(conf)
+    val rdd = sc.makeRDD(1 to 10, 4)
+    val rdd2 = rdd.barrier().mapPartitions { (it, context) =>
+      // Sleep for a random time before global sync.
+      Thread.sleep(Random.nextInt(1000))
+      context.barrier()
+      val time1 = System.currentTimeMillis()
+      // Sleep for a random time between two global syncs.
+      Thread.sleep(Random.nextInt(1000))
+      context.barrier()
+      val time2 = System.currentTimeMillis()
+      Seq((time1, time2)).iterator
+    }
+
+    val times = rdd2.collect()
+    // All the tasks shall finish the first round of global sync within a short time slot.
+    val times1 = times.map(_._1)
+    assert(times1.max - times1.min <= 1000)
+
+    // All the tasks shall finish the second round of global sync within a short time slot.
+    val times2 = times.map(_._2)
+    assert(times2.max - times2.min <= 1000)
+  }
+
+  test("throw exception on barrier() call timeout") {
+    val conf = new SparkConf()
+      .set("spark.barrier.sync.timeout", "1")
+      .set("spark.test.noStageRetry", "true")
+      .setMaster("local-cluster[4, 1, 1024]")
+      .setAppName("test-cluster")
+    sc = new SparkContext(conf)
+    val rdd = sc.makeRDD(1 to 10, 4)
+    val rdd2 = rdd.barrier().mapPartitions { (it, context) =>
+      // Task 3 shall sleep 2000ms to ensure barrier() call timeout
+      if (context.taskAttemptId == 3) {
+        Thread.sleep(2000)
+      }
+      context.barrier()
+      it
+    }
+
+    val error = intercept[SparkException] {
+      rdd2.collect()
+    }.getMessage
+    assert(error.contains("The coordinator didn't get all barrier sync requests"))
+    assert(error.contains("within 1s"))
+  }
+
+  test("throw exception if barrier() call doesn't happen on every task") {
+    val conf = new SparkConf()
+      .set("spark.barrier.sync.timeout", "1")
+      .set("spark.test.noStageRetry", "true")
+      .setMaster("local-cluster[4, 1, 1024]")
+      .setAppName("test-cluster")
+    sc = new SparkContext(conf)
+    val rdd = sc.makeRDD(1 to 10, 4)
+    val rdd2 = rdd.barrier().mapPartitions { (it, context) =>
+      if (context.taskAttemptId != 0) {
+        context.barrier()
+      }
+      it
+    }
+
+    val error = intercept[SparkException] {
+      rdd2.collect()
+    }.getMessage
+    assert(error.contains("The coordinator didn't get all barrier sync requests"))
+    assert(error.contains("within 1s"))
+  }
+
+  test("throw exception if the number of barrier() calls are not the same on every task") {
+    val conf = new SparkConf()
+      .set("spark.barrier.sync.timeout", "1")
+      .set("spark.test.noStageRetry", "true")
+      .setMaster("local-cluster[4, 1, 1024]")
+      .setAppName("test-cluster")
+    sc = new SparkContext(conf)
+    val rdd = sc.makeRDD(1 to 10, 4)
+    val rdd2 = rdd.barrier().mapPartitions { (it, context) =>
+      try {
```

Comment:
seems can be simpler
```
if (context.taskAttemptId == 0) {
  context.barrier()
  context.barrier()
} else {
  context.barrier()
}
```

Reply:
This actually shows one kind of wrong use case that accidentally skipped a `barrier()` call for a task. I updated the comment to show why we issue the problem in a less straight forward way.